store/db.go

```diff
@@ -49,15 +51,17 @@ func newDBLock(dirPath string) (*dbLock, error) {
 }
 
 func (dl *dbLock) lock() error {
+	dl.Lock()
 	if err := dl.fl.ExclusiveLock(); err != nil {
+		dl.Unlock()
 		return err
 	}
-	dl.Lock()
 	return nil
 }
 
 func (dl *dbLock) unlock() error {
 	if err := dl.fl.Unlock(); err != nil {
+		dl.Unlock()
```

Comment:
I think you can end up in a bad state here.

If the `Flock` call fails, it's very possible that the file lock is still held, and yet you release the guarding mutex here. the `lock` function will now never succeed (returning an error each time).

However, it's worse than that, since the caller can't fix this state. If the caller doesn't call `unlock` again, they're broken in that `lock` will never work again. However, if they call `unlock` again to try and fix this, they'll end up hitting `dl.Unlock` with it already being unlocked, which is a panic.

I think the correct fix is to keep the above `dl.Lock` move up, but to only `Unlock` here if the `fl.Unlock` succeeds, and then have a caller up the line retry.

There's a couple additional subtleties in that an error from `fl.Unlock` does not necessarily mean that the lock can ever be unlocked, so you really need to inspect the error and behave differently depending.

It's possible I'm missing something here and the above scenario can't happen.